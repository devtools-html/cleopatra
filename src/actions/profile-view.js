/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import type {
  ProfileSelection, CallTreeFilter, ImplementationFilter,
} from '../types/actions';
import type { Action, ThunkAction } from '../types/store';
import type { ThreadIndex, IndexIntoFuncTable, IndexIntoMarkersTable } from '../types/profile';

/**
 * The actions that pertain to changing the view on the profile, including searching
 * and filtering. Currently the call tree's actions are in this file, but should be
 * split apart. These actions should most likely affect every panel.
 */
export function changeSelectedFuncStack(
  threadIndex: ThreadIndex,
  selectedFuncStack: IndexIntoFuncTable[]
): Action {
  return {
    type: 'CHANGE_SELECTED_FUNC_STACK',
    selectedFuncStack, threadIndex,
  };
}

export function changeSelectedThread(selectedThread: ThreadIndex): Action {
  return {
    type: 'CHANGE_SELECTED_THREAD',
    selectedThread,
  };
}

export function changeThreadOrder(threadOrder: ThreadIndex[]): Action {
  return {
    type: 'CHANGE_THREAD_ORDER',
    threadOrder,
  };
}

export function hideThread(
  threadIndex: ThreadIndex,
  threadOrder: ThreadIndex[],
  hiddenThreads: ThreadIndex[]
): ThunkAction<void> {
  return dispatch => {
    // Do not allow hiding the last thread.
    if (hiddenThreads.length + 1 === threadOrder.length) {
      return;
    }

    dispatch(({
      type: 'HIDE_THREAD',
      threadIndex,
      threadOrder,
      hiddenThreads,
    }: Action));
  };
}

export function showThread(threadIndex: ThreadIndex): Action {
  return {
    type: 'SHOW_THREAD',
    threadIndex,
  };
}

export function changeCallTreeSearchString(searchString: string): Action {
  return {
    type: 'CHANGE_CALL_TREE_SEARCH_STRING',
    searchString,
  };
}

export function changeExpandedFuncStacks(
  threadIndex: ThreadIndex,
  expandedFuncStacks: Array<IndexIntoFuncTable[]>
): Action {
  return {
    type: 'CHANGE_EXPANDED_FUNC_STACKS',
    threadIndex, expandedFuncStacks,
  };
}

export function changeSelectedMarker(
  threadIndex: ThreadIndex,
  selectedMarker: IndexIntoMarkersTable | -1
): Action {
  return {
    type: 'CHANGE_SELECTED_MARKER',
    selectedMarker, threadIndex,
  };
}

export function changeImplementationFilter(implementation: ImplementationFilter): Action {
  return {
    type: 'CHANGE_IMPLEMENTATION_FILTER',
    implementation,
  };
}

export function changeInvertCallstack(invertCallstack: boolean): Action {
  return {
    type: 'CHANGE_INVERT_CALLSTACK',
    invertCallstack,
  };
}

export function changeHidePlatformDetails(hidePlatformDetails: boolean): Action {
  return {
    type: 'CHANGE_HIDE_PLATFORM_DETAILS',
    hidePlatformDetails,
  };
}

export type UpdateProfileSelection = (selection: ProfileSelection) => Action;

export function updateProfileSelection(selection: ProfileSelection): Action {
  return {
    type: 'UPDATE_PROFILE_SELECTION',
    selection,
  };
}

export function addRangeFilter(start: number, end: number): Action {
  return {
    type: 'ADD_RANGE_FILTER',
    start, end,
  };
}

export function addRangeFilterAndUnsetSelection(start: number, end: number): ThunkAction<void> {
  return dispatch => {
    dispatch(addRangeFilter(start, end));
    dispatch(updateProfileSelection({ hasSelection: false, isModifying: false }));
  };
}

export function popRangeFilters(firstRemovedFilterIndex: number): Action {
  return {
    type: 'POP_RANGE_FILTERS',
    firstRemovedFilterIndex,
  };
}

export function popRangeFiltersAndUnsetSelection(firstRemovedFilterIndex: number): ThunkAction<void> {
  return dispatch => {
    dispatch(popRangeFilters(firstRemovedFilterIndex));
    dispatch(updateProfileSelection({ hasSelection: false, isModifying: false }));
  };
}

export function addCallTreeFilter(threadIndex: ThreadIndex, filter: CallTreeFilter): Action {
  return {
    type: 'ADD_CALL_TREE_FILTER',
    threadIndex,
    filter,
  };
}

export function popCallTreeFilters(threadIndex: ThreadIndex, firstRemovedFilterIndex: number): Action {
  return {
    type: 'POP_CALL_TREE_FILTERS',
    threadIndex,
    firstRemovedFilterIndex,
  };
}
