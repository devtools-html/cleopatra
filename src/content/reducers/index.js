import profileView from './profile-view';
import app from './app';
import summaryView from './summary-view';
import urlState from './url-state';
import flameChart from './flame-chart';
import timelineView from './timeline-view';

export default { app, profileView, summaryView, urlState, flameChart, timelineView };
