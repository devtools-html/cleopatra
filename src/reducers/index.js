/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import profileView from './profile-view';
import app from './app';
import summaryView from './summary-view';
import urlState from './url-state';
import flameChart from './flame-chart';
import timelineView from './timeline-view';
import icons from './icons';
import { combineReducers } from 'redux';

export default combineReducers({ app, profileView, summaryView, urlState, flameChart, timelineView, icons });
