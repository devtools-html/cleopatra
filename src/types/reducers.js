/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

import type {
  Action,
  DataSource,
  ProfileSelection,
  ImplementationFilter,
  TabSlug,
} from './actions';
import type { Milliseconds, StartEndRange } from './units';
import type { IndexIntoMarkersTable, Profile, ThreadIndex } from './profile';
import type { CallNodePath } from './profile-derived';
import type { Attempt } from '../utils/errors';
import type { GetLabel } from '../profile-logic/labeling-strategies';
import type { GetCategory } from '../profile-logic/color-categories';
import type { TransformStacksPerThread } from './transforms';
import type JSZip from 'jszip';
import type { IndexIntoZipFileTable } from '../profile-logic/zip-files';

export type Reducer<T> = (T, Action) => T;

export type RequestedLib = { debugName: string, breakpadId: string };
export type SymbolicationStatus = 'DONE' | 'SYMBOLICATING';
export type ThreadViewOptions = {
  selectedCallNodePath: CallNodePath,
  expandedCallNodePaths: Array<CallNodePath>,
  selectedMarker: IndexIntoMarkersTable | -1,
};
export type ProfileViewState = {
  viewOptions: {
    perThread: ThreadViewOptions[],
    symbolicationStatus: SymbolicationStatus,
    waitingForLibs: Set<RequestedLib>,
    selection: ProfileSelection,
    scrollToSelectionGeneration: number,
    focusCallTreeGeneration: number,
    rootRange: StartEndRange,
    zeroAt: Milliseconds,
    tabOrder: number[],
    rightClickedThread: ThreadIndex,
    isCallNodeContextMenuVisible: boolean,
  },
  profile: Profile | null,
};

export type AppViewState =
  | {| +phase: 'ROUTE_NOT_FOUND' |}
  | {| +phase: 'DATA_LOADED' |}
  | {| +phase: 'FATAL_ERROR', +error: Error |}
  | {|
      +phase: 'INITIALIZING',
      +additionalData?: {| +attempt: Attempt | null, +message: string |},
    |};

/**
 * This represents the finite state machine for loading zip files. The phase represents
 * where the state is now.
 */
export type ZipFileState =
  | {|
      +phase: 'NO_ZIP_FILE',
      +zip: null,
      +pathInZipFile: null,
    |}
  | {|
      +phase: 'LIST_FILES_IN_ZIP_FILE',
      +zip: JSZip,
      +pathInZipFile: null,
    |}
  | {|
      +phase: 'PROCESS_PROFILE_FROM_ZIP_FILE',
      +zip: JSZip,
      +pathInZipFile: string,
    |}
  | {|
      +phase: 'FAILED_TO_PROCESS_PROFILE_FROM_ZIP_FILE',
      +zip: JSZip,
      +pathInZipFile: string,
    |}
  | {|
      +phase: 'FILE_NOT_FOUND_IN_ZIP_FILE',
      +zip: JSZip,
      +pathInZipFile: string,
    |}
  | {|
      +phase: 'VIEW_PROFILE_IN_ZIP_FILE',
      +zip: JSZip,
      +pathInZipFile: string,
    |};

export type AppState = {
  view: AppViewState,
  isUrlSetupDone: boolean,
  hasZoomedViaMousewheel: boolean,
};

export type ZippedProfilesState = {
  zipFile: ZipFileState,
  selectedZipFileIndex: IndexIntoZipFileTable | null,
  // In practice this should never contain null, but needs to support the
  // TreeView interface.
  expandedZipFileIndexes: Array<IndexIntoZipFileTable | null>,
};

export type RangeFilterState = {
  start: number,
  end: number,
};

export type UrlState = {
  dataSource: DataSource,
  hash: string,
  profileUrl: string,
  selectedTab: TabSlug,
  rangeFilters: RangeFilterState[],
  selectedThread: ThreadIndex | null,
  callTreeSearchString: string,
  markersSearchString: string,
  implementation: ImplementationFilter,
  invertCallstack: boolean,
  threadOrder: ThreadIndex[],
  hiddenThreads: ThreadIndex[],
  transforms: TransformStacksPerThread,
  pathInZipFile: string | null,
};

export type IconState = Set<string>;

export type StackChartState = {
  categoryColorStrategy: GetCategory,
  labelingStrategy: GetLabel,
};

export type State = {|
  app: AppState,
  profileView: ProfileViewState,
  urlState: UrlState,
  stackChart: StackChartState,
  icons: IconState,
  zippedProfiles: ZippedProfilesState,
|};

export type IconWithClassName = {
  icon: string,
  className: string,
};
