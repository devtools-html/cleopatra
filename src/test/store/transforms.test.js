/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import {
  getProfileForUnfilteredCallTree,
  getProfileForInvertedCallTree,
  getProfileWithMixedJSImplementation,
} from '../fixtures/profiles/profiles-for-call-trees';

import { storeWithProfile } from '../fixtures/stores';
import {
  addTransformToStack,
  popTransformsFromStack,
  changeInvertCallstack,
  changeImplementationFilter,
  changeSelectedCallNode,
} from '../../actions/profile-view';
import { selectedThreadSelectors } from '../../reducers/profile-view';
import { CallTree } from '../../profile-logic/call-tree';
import type { IndexIntoCallNodeTable } from '../../types/profile-derived';

export function formatTree(
  callTree: CallTree,
  children: IndexIntoCallNodeTable[] = callTree.getRoots(),
  depth: number = 0,
  previousString: string = ''
) {
  const whitespace = Array(depth * 2).join(' ');

  return children.reduce((string, callNodeIndex) => {
    const { name, totalTime, selfTime } = callTree.getNode(callNodeIndex);
    const text = `\n${whitespace}- ${name} (total: ${totalTime}, self:${selfTime})`;
    return formatTree(
      callTree,
      callTree.getChildren(callNodeIndex),
      depth + 1,
      string + text
    );
  }, previousString);
}

describe('"focus-subtree" transform', function() {
  describe('on a call tree', function() {
    /**
     * Assert this transformation:
     *
     *                     A:3,0                              C:2,0
     *                       |                               /      \
     *                       v       Focus [A, B, C]        v        v
     *                     B:3,0           -->           D:1,0     F:1,0
     *                     /    \                         |           |
     *                    v      v                        v           v
     *                C:2,0     H:1,0                   E:1,1       G:1,1
     *               /      \         \
     *              v        v         v
     *            D:1,0     F:1,0     I:1,1
     *            |           |
     *            v           v
     *          E:1,1       G:1,1
     */
    const profile = getProfileForUnfilteredCallTree();
    const { dispatch, getState } = storeWithProfile(profile);
    const originalCallTree = selectedThreadSelectors.getCallTree(getState());
    const threadIndex = 0;
    const A = 0;
    const B = 1;
    const C = 2;

    it('starts as an unfiltered call tree', function() {
      expect(formatTree(originalCallTree)).toMatchSnapshot();
    });

    it('can be filtered to a subtree', function() {
      dispatch(
        addTransformToStack(threadIndex, {
          type: 'focus-subtree',
          callNodePath: [A, B, C],
          implementation: 'combined',
          inverted: false,
        })
      );
      const callTree = selectedThreadSelectors.getCallTree(getState());
      expect(formatTree(callTree)).toMatchSnapshot();
    });

    it('can remove the transform', function() {
      dispatch(popTransformsFromStack(threadIndex, 0));
      const callTree = selectedThreadSelectors.getCallTree(getState());
      const formattedTree = formatTree(callTree);
      expect(formattedTree).toMatchSnapshot();
      expect(formattedTree).toEqual(formatTree(originalCallTree));
    });
  });

  describe('on an inverted call tree', function() {
    /**
     *           1. Starting call tree         ->       2. Invert call tree        ->
     *
     *                  A:3,0                             Z:2,2         E:1,1
     *                    ↓                                 ↓             ↓
     *                  B:3,0                             Y:2,0         D:1,0
     *                 ↙     ↘                              ↓             ↓
     *             X:1,0     C:1,0                        X:2,0         C:1,0
     *            ↙         ↙      ↘                     ↙     ↘          ↓
     *         Y:1,0     X:1,0     D:1,0             B:1,0     C:1,0    B:1,0
     *           ↓         ↓         ↓                 ↓         ↓        ↓
     *        Z:1,1      Y:1,0     E:1,0             A:1,0     B:1,0    A:1,0
     *                     ↓                                     ↓
     *                   Z:1,1                                  A:1,0
     *
     *  --------------------------------------------------------------------------------
     *
     *   ->    3. Focus [Z, Y, X]      ->      4. Un-invert call tree
     *
     *                 X:2,2                           A:2,0
     *                ↙     ↘                            ↓
     *            B:1,0     C:1,0                      B:2,0
     *              ↓         ↓                       ↙     ↘
     *            A:1,0     B:1,0                 X:1,1     C:1,0
     *                        ↓                               ↓
     *                      A:1,0                           X:1,1
     */
    const profile = getProfileForInvertedCallTree();
    const { dispatch, getState } = storeWithProfile(profile);
    dispatch(changeInvertCallstack(true));

    it('starts as an inverted call tree', function() {
      const callTree = selectedThreadSelectors.getCallTree(getState());
      expect(formatTree(callTree)).toMatchSnapshot();
    });

    it('can be filtered to a subtree', function() {
      const threadIndex = 0;
      const X = 5;
      const Y = 6;
      const Z = 7;

      dispatch(
        addTransformToStack(threadIndex, {
          type: 'focus-subtree',
          callNodePath: [Z, Y, X],
          implementation: 'combined',
          inverted: true,
        })
      );
      const callTree = selectedThreadSelectors.getCallTree(getState());
      expect(formatTree(callTree)).toMatchSnapshot();
    });

    it('can be un-inverted and keep the transform', function() {
      dispatch(changeInvertCallstack(false));
      const callTree = selectedThreadSelectors.getCallTree(getState());
      expect(formatTree(callTree)).toMatchSnapshot();
    });
  });
});

describe('"merge-call-node" transform', function() {
  describe('on a call tree', function() {
    /**
     * Assert this transformation:
     *
     *                     A:3,0                              A:3,0
     *                       |                                  |
     *                       v       Focus [A, B, C]            v
     *                     B:3,0           -->                B:3,0
     *                     /    \                           /   |   \
     *                    v      v                         v    v    v
     *                C:2,0     H:1,0                 D:1,0   F:1,0   I:1,1
     *               /      \         \                 |       |
     *              v        v         v                v       v
     *            D:1,0     F:1,0     I:1,1           E:1,1   G:1,1
     *            |           |
     *            v           v
     *          E:1,1       G:1,1
     */
    const profile = getProfileForUnfilteredCallTree();
    const { dispatch, getState } = storeWithProfile(profile);
    const originalCallTree = selectedThreadSelectors.getCallTree(getState());
    const threadIndex = 0;
    const A = 0;
    const B = 1;
    const C = 2;

    it('starts as an unfiltered call tree', function() {
      expect(formatTree(originalCallTree)).toMatchSnapshot();
    });

    it('call node [A, B, C] can be merged into [A, B]', function() {
      dispatch(
        addTransformToStack(threadIndex, {
          type: 'merge-call-node',
          callNodePath: [A, B, C],
          implementation: 'combined',
          inverted: false,
        })
      );
      const callTree = selectedThreadSelectors.getCallTree(getState());
      expect(formatTree(callTree)).toMatchSnapshot();
    });
  });

  describe('on a JS call tree', function() {
    const profile = getProfileWithMixedJSImplementation();
    const threadIndex = 0;

    // funcIndexes in the profile fixture.
    /* eslint-disable no-unused-vars */
    const RUN_SCRIPT = 0;
    const ON_LOAD = 1;
    const A = 2;
    const B = 3;
    const ION_CANNON = 4;
    /* eslint-enable no-unused-vars */

    const mergeJSPathAB = {
      type: 'merge-call-node',
      callNodePath: [ON_LOAD, A],
      implementation: 'js',
      inverted: false,
    };

    const mergeCombinedPathToA = {
      type: 'merge-call-node',
      callNodePath: [RUN_SCRIPT, ON_LOAD, A],
      implementation: 'combined',
      inverted: false,
    };

    it('starts as an untransformed call tree', function() {
      /**
       *     JS::RunScript
       *          ↓
       *        onLoad
       *      ↙       ↘
       *     a        js::jit::IonCannon
       *     ↓             ↓
       *     b             a
       *                   ↓
       *                   b
       */
      const { getState } = storeWithProfile(profile);
      expect(
        formatTree(selectedThreadSelectors.getCallTree(getState()))
      ).toMatchSnapshot();
    });

    it('has an untransformed JS only view', function() {
      /**
       *    onLoad
       *      ↓
       *      a
       *      ↓
       *      b
       */
      const { dispatch, getState } = storeWithProfile(profile);
      dispatch(changeImplementationFilter('js'));
      expect(
        formatTree(selectedThreadSelectors.getCallTree(getState()))
      ).toMatchSnapshot();
    });

    it('can merge the node at JS path "onLoad" -> "A"', function() {
      /**
       *    onLoad
       *      ↓
       *      b
       */
      const { dispatch, getState } = storeWithProfile(profile);
      dispatch(changeImplementationFilter('js'));
      dispatch(addTransformToStack(threadIndex, mergeJSPathAB));
      expect(
        formatTree(selectedThreadSelectors.getCallTree(getState()))
      ).toMatchSnapshot();
    });

    it('can merge the node at JS path "onLoad" -> "A" on an combined call tree', function() {
      /**
       *     JS::RunScript
       *          ↓
       *        onLoad
       *      ↙       ↘
       *     b        js::jit::IonCannon
       *                   ↓
       *                   b
       */
      const { dispatch, getState } = storeWithProfile(profile);
      dispatch(addTransformToStack(threadIndex, mergeJSPathAB));
      expect(
        formatTree(selectedThreadSelectors.getCallTree(getState()))
      ).toMatchSnapshot();
    });

    it('can merge a combined CallNodePath, and display a correct JS call tree', function() {
      /**
       *    onLoad
       *   ↙      ↘
       *  b        a
       *           ↓
       *           b
       */
      const { dispatch, getState } = storeWithProfile(profile);
      dispatch(changeImplementationFilter('js'));
      dispatch(addTransformToStack(threadIndex, mergeCombinedPathToA));
      expect(
        formatTree(selectedThreadSelectors.getCallTree(getState()))
      ).toMatchSnapshot();
    });

    it('starts as an inverted call tree', function() {
      /**
       *                       b
       *                       ↓
       *                       a
       *                   ↙       ↘
       *  js::jit::IonCannon      onLoad
       *          ↓                 ↓
       *       onLoad          JS::RunScript
       *          ↓
       *    JS::RunScript
       */
      const { dispatch, getState } = storeWithProfile(profile);
      dispatch(changeInvertCallstack(true));
      expect(
        formatTree(selectedThreadSelectors.getCallTree(getState()))
      ).toMatchSnapshot();
    });

    it('can merge path [b, a, js::jit::IonCannon] on an inverted call tree', function() {
      /**
       *          b
       *          ↓
       *          a
       *          ↓
       *        onLoad
       *          ↓
       *    JS::RunScript
       */
      const { dispatch, getState } = storeWithProfile(profile);
      dispatch(changeInvertCallstack(true));
      dispatch(
        addTransformToStack(threadIndex, {
          type: 'merge-call-node',
          callNodePath: [B, A, ION_CANNON],
          implementation: 'combined',
          inverted: true,
        })
      );

      expect(
        formatTree(selectedThreadSelectors.getCallTree(getState()))
      ).toMatchSnapshot();
    });

    it('can merge path [b, a, onLoad] on an inverted JS call tree', function() {
      /**
       *                       b
       *                       ↓
       *                       a
       *                   ↙       ↘
       *  js::jit::IonCannon       JS::RunScript
       *           ↓
       *    JS::RunScript
       */
      const { dispatch, getState } = storeWithProfile(profile);
      dispatch(changeInvertCallstack(true));
      dispatch(
        addTransformToStack(threadIndex, {
          type: 'merge-call-node',
          callNodePath: [B, A, ON_LOAD],
          implementation: 'js',
          inverted: true,
        })
      );

      expect(
        formatTree(selectedThreadSelectors.getCallTree(getState()))
      ).toMatchSnapshot();
    });
  });
});

describe('expanded and selected CallNodePaths', function() {
  const profile = getProfileForUnfilteredCallTree();
  const threadIndex = 0;
  const A = 0;
  const B = 1;
  const C = 2;
  const D = 3;
  const selectedCallNodePath = [A, B, C, D];

  it('can select a path and expand the nodes to that path', function() {
    const { dispatch, getState } = storeWithProfile(profile);
    // This opens expands the call nodes up to this point.
    dispatch(changeSelectedCallNode(threadIndex, selectedCallNodePath));
    expect(
      selectedThreadSelectors.getSelectedCallNodePath(getState())
    ).toEqual([A, B, C, D]);

    expect(
      selectedThreadSelectors.getExpandedCallNodePaths(getState())
    ).toEqual([
      // Expanded nodes:
      [A],
      [A, B],
      [A, B, C],
    ]);
  });

  it('can update call node references for focusing a subtree', function() {
    const { dispatch, getState } = storeWithProfile(profile);
    // This opens expands the call nodes up to this point.
    dispatch(changeSelectedCallNode(threadIndex, [A, B, C, D]));
    dispatch(
      addTransformToStack(threadIndex, {
        type: 'focus-subtree',
        callNodePath: [A, B],
        implementation: 'combined',
        inverted: false,
      })
    );

    expect(
      selectedThreadSelectors.getSelectedCallNodePath(getState())
    ).toEqual([B, C, D]);
    expect(
      selectedThreadSelectors.getExpandedCallNodePaths(getState())
    ).toEqual([
      // Expanded nodes:
      [B],
      [B, C],
    ]);
  });

  it('can update call node references for merging a node', function() {
    const { dispatch, getState } = storeWithProfile(profile);
    // This opens expands the call nodes up to this point.
    dispatch(changeSelectedCallNode(threadIndex, [A, B, C, D]));
    dispatch(
      addTransformToStack(threadIndex, {
        type: 'merge-call-node',
        callNodePath: [A, B],
        implementation: 'combined',
        inverted: false,
      })
    );

    expect(
      selectedThreadSelectors.getSelectedCallNodePath(getState())
    ).toEqual([A, C, D]);
    expect(
      selectedThreadSelectors.getExpandedCallNodePaths(getState())
    ).toEqual([
      // Expanded nodes:
      [A],
      [A, C],
    ]);
  });
});

describe('expanded and selected CallNodePaths on inverted trees', function() {
  const profile = getProfileForUnfilteredCallTree();
  const threadIndex = 0;
  const B = 1;
  const X = 5;
  const Y = 6;
  const Z = 7;

  const selectedCallNodePath = [Z, Y, X, B];

  it('can select an inverted path and expand the nodes to that path', function() {
    const { dispatch, getState } = storeWithProfile(profile);
    // This opens expands the call nodes up to this point.
    dispatch(changeInvertCallstack(true));
    dispatch(changeSelectedCallNode(threadIndex, selectedCallNodePath));

    expect(
      selectedThreadSelectors.getSelectedCallNodePath(getState())
    ).toEqual([Z, Y, X, B]);
    expect(
      selectedThreadSelectors.getExpandedCallNodePaths(getState())
    ).toEqual([
      // Expanded nodes:
      [Z],
      [Z, Y],
      [Z, Y, X],
    ]);
  });

  it('can update call node references for focusing a subtree', function() {
    const { dispatch, getState } = storeWithProfile(profile);
    // This opens expands the call nodes up to this point.
    dispatch(changeSelectedCallNode(threadIndex, selectedCallNodePath));
    dispatch(changeInvertCallstack(true));
    dispatch(
      addTransformToStack(threadIndex, {
        type: 'focus-subtree',
        callNodePath: [Z, Y],
        implementation: 'combined',
        inverted: false,
      })
    );

    expect(
      selectedThreadSelectors.getSelectedCallNodePath(getState())
    ).toEqual([Y, X, B]);
    expect(
      selectedThreadSelectors.getExpandedCallNodePaths(getState())
    ).toEqual([
      // Expanded nodes:
      [Y],
      [Y, X],
    ]);
  });

  it('can update call node references for merging a call node', function() {
    const { dispatch, getState } = storeWithProfile(profile);
    // This opens expands the call nodes up to this point.
    dispatch(changeSelectedCallNode(threadIndex, selectedCallNodePath));
    dispatch(changeInvertCallstack(true));
    dispatch(
      addTransformToStack(threadIndex, {
        type: 'merge-call-node',
        callNodePath: [Z, Y],
        implementation: 'combined',
        inverted: false,
      })
    );

    expect(
      selectedThreadSelectors.getSelectedCallNodePath(getState())
    ).toEqual([Z, X, B]);
    expect(
      selectedThreadSelectors.getExpandedCallNodePaths(getState())
    ).toEqual([
      // Expanded nodes:
      [Z],
      [Z, X],
    ]);
  });
});
