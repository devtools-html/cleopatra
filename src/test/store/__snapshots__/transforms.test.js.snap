// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`"focus-subtree" transform on a call tree can be filtered to a subtree 1`] = `
"
- C (total: 2, self:—)
 - D (total: 1, self:—)
   - E (total: 1, self:1)
 - F (total: 1, self:—)
   - G (total: 1, self:1)"
`;

exports[`"focus-subtree" transform on a call tree can remove the transform 1`] = `
"
- A (total: 3, self:—)
 - B (total: 3, self:—)
   - C (total: 2, self:—)
     - D (total: 1, self:—)
       - E (total: 1, self:1)
     - F (total: 1, self:—)
       - G (total: 1, self:1)
   - H (total: 1, self:—)
     - I (total: 1, self:1)"
`;

exports[`"focus-subtree" transform on a call tree starts as an unfiltered call tree 1`] = `
"
- A (total: 3, self:—)
 - B (total: 3, self:—)
   - C (total: 2, self:—)
     - D (total: 1, self:—)
       - E (total: 1, self:1)
     - F (total: 1, self:—)
       - G (total: 1, self:1)
   - H (total: 1, self:—)
     - I (total: 1, self:1)"
`;

exports[`"focus-subtree" transform on an inverted call tree can be filtered to a subtree 1`] = `
"
- X (total: 2, self:2)
 - B (total: 1, self:—)
   - A (total: 1, self:—)
 - C (total: 1, self:—)
   - B (total: 1, self:—)
     - A (total: 1, self:—)"
`;

exports[`"focus-subtree" transform on an inverted call tree can be un-inverted and keep the transform 1`] = `
"
- A (total: 2, self:—)
 - B (total: 2, self:—)
   - C (total: 1, self:—)
     - X (total: 1, self:1)
   - X (total: 1, self:1)"
`;

exports[`"focus-subtree" transform on an inverted call tree starts as an inverted call tree 1`] = `
"
- Z (total: 2, self:2)
 - Y (total: 2, self:—)
   - X (total: 2, self:—)
     - B (total: 1, self:—)
       - A (total: 1, self:—)
     - C (total: 1, self:—)
       - B (total: 1, self:—)
         - A (total: 1, self:—)
- E (total: 1, self:1)
 - D (total: 1, self:—)
   - C (total: 1, self:—)
     - B (total: 1, self:—)
       - A (total: 1, self:—)"
`;

exports[`"merge-call-node" transform on a JS call tree can merge a combined CallNodePath, and display a correct JS call tree 1`] = `
"
- onLoad (total: 3, self:—)
 - a (total: 2, self:—)
   - b (total: 2, self:2)
 - b (total: 1, self:1)"
`;

exports[`"merge-call-node" transform on a JS call tree can merge path [b, a, js::jit::IonCannon] on an inverted call tree 1`] = `
"
- b (total: 3, self:3)
 - a (total: 3, self:—)
   - onLoad (total: 3, self:—)
     - JS::RunScript (total: 3, self:—)"
`;

exports[`"merge-call-node" transform on a JS call tree can merge path [b, a, onLoad] on an inverted JS call tree 1`] = `
"
- b (total: 3, self:3)
 - a (total: 3, self:—)
   - js::jit::IonCannon (total: 2, self:—)
     - JS::RunScript (total: 2, self:—)
   - JS::RunScript (total: 1, self:—)"
`;

exports[`"merge-call-node" transform on a JS call tree can merge the node at JS path "onLoad" -> "A" 1`] = `
"
- onLoad (total: 3, self:—)
 - b (total: 3, self:3)"
`;

exports[`"merge-call-node" transform on a JS call tree can merge the node at JS path "onLoad" -> "A" on an combined call tree 1`] = `
"
- JS::RunScript (total: 3, self:—)
 - onLoad (total: 3, self:—)
   - js::jit::IonCannon (total: 2, self:—)
     - b (total: 2, self:2)
   - b (total: 1, self:1)"
`;

exports[`"merge-call-node" transform on a JS call tree has an untransformed JS only view 1`] = `
"
- onLoad (total: 3, self:—)
 - a (total: 3, self:—)
   - b (total: 3, self:3)"
`;

exports[`"merge-call-node" transform on a JS call tree starts as an inverted call tree 1`] = `
"
- b (total: 3, self:3)
 - a (total: 3, self:—)
   - js::jit::IonCannon (total: 2, self:—)
     - onLoad (total: 2, self:—)
       - JS::RunScript (total: 2, self:—)
   - onLoad (total: 1, self:—)
     - JS::RunScript (total: 1, self:—)"
`;

exports[`"merge-call-node" transform on a JS call tree starts as an untransformed call tree 1`] = `
"
- JS::RunScript (total: 3, self:—)
 - onLoad (total: 3, self:—)
   - js::jit::IonCannon (total: 2, self:—)
     - a (total: 2, self:—)
       - b (total: 2, self:2)
   - a (total: 1, self:—)
     - b (total: 1, self:1)"
`;

exports[`"merge-call-node" transform on a call tree call node [A, B, C] can be merged into [A, B] 1`] = `
"
- A (total: 3, self:—)
 - B (total: 3, self:—)
   - D (total: 1, self:—)
     - E (total: 1, self:1)
   - F (total: 1, self:—)
     - G (total: 1, self:1)
   - H (total: 1, self:—)
     - I (total: 1, self:1)"
`;

exports[`"merge-call-node" transform on a call tree starts as an unfiltered call tree 1`] = `
"
- A (total: 3, self:—)
 - B (total: 3, self:—)
   - C (total: 2, self:—)
     - D (total: 1, self:—)
       - E (total: 1, self:1)
     - F (total: 1, self:—)
       - G (total: 1, self:1)
   - H (total: 1, self:—)
     - I (total: 1, self:1)"
`;
